### PR TITLE
Porting roslyn regex tests to corefx

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/corefx/issues/28508
Fixes https://github.com/dotnet/corefx/issues/27618

- Adding internal `RegexParseException` that derives from ArgumentException to provide internal error information `RegexParseError`.
- Porting test cases from the roslyn PR over to corefx and changing them to InlineData instead of a method per test case.
- Asserting on the specific `RegexParseError` for negative tests to be more precise (existing tests + roslyn tests)

TODO:
- [x] Adding sub expression checks (probably without asserting on the specific RegexParseError)
- [x] Refine RegexParseError options and add comments to each.

### CODE REVIEW

File: src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs]: We've tried to avoid doing this,  preferring reflection if necessary instead. Other issues aside, now that this will prevent the Linker from shaking away dead code.
cc: @weshaggard

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs]: I wouldn't be able to use attribute based testing if the `RegexParseError` enum isn't public: https://github.com/dotnet/corefx/pull/29178/files#diff-e55ccc955dc3ddac57a0e77d1a8fc944R947 and I wouldn't be able to catch on a the specific exception. I could workaround the second issue easily by catching the parent exception and then checking the exact type via reflection and its error code but I'm not sure how to handle the first issue. We are speaking about ~1000 test cases here, switching to strings or ints instead of the enum value doesn't seem right?

Comment (weshaggard) [reply src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs]: Yes lets please try and avoid InternalsVisibleTo as it causes coupling that is too tight to one particular implementation.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 2]
```diff
@@ -420,7 +423,7 @@ private RegexNode ScanRegex()
                             break;
 
                         default:
-                            throw MakeException(SR.InternalError);
+                            throw MakeException(RegexParseError.InternalError, SR.InternalError);
```
Comment (CyrusNajmabadi) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: this doesn't seem like this should be a parse/argument exception.  this should just be something like an InvalidOperationException.

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 3]
```diff
@@ -0,0 +1,1029 @@
...
+        [InlineData(@"@""\1""", RegexOptions.None, RegexParseError.UndefinedBackref)]
+        [InlineData(@"@""(?')""", RegexOptions.None, RegexParseError.InvalidGroupName)]
+        [InlineData(@"@""(?<)""", RegexOptions.None, RegexParseError.InvalidGroupName)]
+        [InlineData(@"@""(?)""", RegexOptions.None, RegexParseError.QuantifyAfterNothing)]
+        [InlineData(@"@""(?>""", RegexOptions.None, RegexParseError.NotEnoughParens)]
+        [InlineData(@"@""(?<!""", RegexOptions.None, RegexParseError.NotEnoughParens)]
+        [InlineData(@"@""(?<=""", RegexOptions.None, RegexParseError.NotEnoughParens)]
+        [InlineData(@"@""(?!""", RegexOptions.None, RegexParseError.NotEnoughParens)]
+        [InlineData(@"@""(?=""", RegexOptions.None, RegexParseError.NotEnoughParens)]
+        [InlineData(@"@""(?imn )""", RegexOptions.None, RegexParseError.UnrecognizedGro
```
Comment (CyrusNajmabadi) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: are you intending to implement this?

Comment (CyrusNajmabadi) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: these are probably fine to run with subtree checks, now that @danmosemsft put a limit on the memory allocation for these types of regexes

Comment (CyrusNajmabadi) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: this should be no problem for you guys, i don't tink.  You don't use actual stack recursion when parsing.  This was a special test for Roslyn because my parser is recursive descent, and so i needed to make sure something complex like this bailed out properly without causing a StackOverflowException.

Comment (CyrusNajmabadi) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: this and the 'if' above coudl just be made into a simple assert right?

Comment (danmosemsft) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: what is this?

Comment (danmosemsft) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: Can you add comments explaining what this verification is doing/why

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: Technically you literally should not be able to hit this.  This is an internal error in your code.  It's the difference between improper input from teh user, versus a bug in your own code.

It's no more of a breaking change than the change you made to stop throwing an 'index out of bounds error' based off of the bugs i filed.  Technically someone could have been dependent on internal details about how your code crashes... but i don't think .net ever tries to preserve internal crashing behavior.

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 3 above)
Comment (danmosemsft) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: I guess I'm unfamliar with what subtree means in this context.

Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: SubTree is something i did in the roslyn tests.  specifically, given any input string we mangle it in many different ways and validate that we dont' crash on any of the mangled names.  Examples of how i mangled:

1.  take the starting string and independently remove each character from it.  i.e.  ```"cat" -> ["at", "ct", "ca"]```.  test all those cases.
2. take the starting string and remove larger and larger prefixes/suffixes.  i.e. ```"cat" -> ["at", "t", "ca", "c"]```.

etc. etc.

basically, a minor form of fuzzing for regexes.  Note: actual fuzzing would be highly appropriate here.  Regex parsing is basically the poster child for that sort of an approach.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        NotEnoughParens,
+        IllegalRange,
+        InternalError,
+        NestedQuantify,
+        QuantifyAfterNothing,
+        TooManyParens,
+        UnknownProperty // Unicode block, \p{Property}
+    }
+
+    internal class RegexParseException : ArgumentException
```
[the remark this answers is not in the document]
Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: The owners of the System.Xml parser did not want to include error codes for exactly this reason. Given we're now SxS, the concern about breaks is a little lower. Certainly changing the error code in an exception is much less likely to break folks than changing a success case to an error case. And changing to another existing code a little less likely than changing to a newly added error code. I think it's probably OK.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: Note: it's potentially worth pointing out that Roslyn (which exposes parses and handles tons of input error cases) *specifically* does not expose any type of enum for this sort of thing.  Turns out, in practice, it's not that useful, and it's a bit of a maintenance burden.  

instead, it simply returns back trees along with "diagnostics" which are effectively opaque objects that just say "here's a message about what's wrong, and here's the location it happened".  That's been highly sufficient for our needs (and we specifically expose this stuff so people can analyze code effectively).

This is also the model i took for the new Roslyn regex parser.  

1. outside of stack-overflows, you *always* get a tree back.
2. the tree always contains all input text of the original user input (including whitespace/comments).
3. issues are returned as diagnostics that are provided along with the tree.  diagnostics are simply a message, along with a position/span of relevance.  This is used for doing things like placing a squiggle appropriately.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: > How did you end up doing this -- did you copy and instrument our parser, or write your own?

i rewrote it entirely, using the reference specification, and then diving into your impl to see how things deviated at times.  For example, character class parsing is exceptionally strange in the native parser.  But i replicated it in on our end because compatibility with the core parser is the #1 priority.  Tons of tests were added (i.e. the basis of htis PR) to validate the same observable behavior between the two systems.

Note: i originally tried to just fork and use the native code.  However, it was far too dfificult as the existing code is designed to generate a lossy AST in place, which it continually mutates as it proceeds.  Trying to work off of that was just too difficult.  

When necessary, i also took code wholesale.  For example, the unicode handling code was copied practically verbatim here: https://github.com/dotnet/roslyn/pull/23984/files#diff-f689c52f48c471beeaea4252031ed8f3

**Comment by danmosemsft** [comment]
I imagine you'll have a small number you'll need to disable for NETFX runs (since we fixed a few parser bugs) but otherwise I hope to see all these pass on both NETFX and Core.

File: src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs (hunk 1 above)
Comment (akoeplinger) [reply src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs]: Using reflection instead of IVT also makes it easier for us in Mono to consume the tests since our test assemblies have different names 😄

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/Properties/InternalsVisibleTo.cs]: Great, thanks for the input.

File: src/System.Text.RegularExpressions/tests/Regex.Ctor.Tests.cs [hunk 5]
```diff
@@ -99,197 +99,5 @@ public static void StaticCtor_InvalidTimeoutRange_ExceptionThrown()
                 return SuccessExitCode;
             }).Dispose();
         }
-
-        [Theory]
-        // \d, \D, \s, \S, \w, \W, \P, \p inside character range
-        [InlineData(@"cat([a-\d]*)dog", RegexOptions.None)]
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/Regex.Ctor.Tests.cs]: jep, I ported/moved all the tests directly to RegexParserTetsts. Nothing should have slipped out of my radar.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: > It seems like you're signign yourself up to very specific ways of parsing in order to preserve the error codes you emit.

Given it's a field on an internal type, I do think we would be OK with letting this value change with parser modifications. If the field is public I think it's something to think about.

[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: The point why I added the enum in the first place was that I wanted to assert that specific patterns cause specific errors. With that I can be more precise and catch only cases that I expect to fail. I noticed that in your test bed you catched on various error cases like ArgumentException, InvalidOperationException, ... globally.

**Comment by danmosemsft** [comment]
> I notice that it seems important to you to run these tests also on netfx so I will add some extra code to handle that scenario.

Ha, yes you're right: reason is that our regex code has diverged significantly from netfx, and netfx is the the only oracle that we can use to verify we aren't going off course. That is the fixed point that makes it safe for us to do the kind of refactoring you're doing.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: I ran all the existing tests + the roslyn ones and never hit the internal error. I agree with @CyrusNajmabadi that changing to InvalidOperationException make more sense here. Are you ok with that Dan?

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: sorry which code line are referring to? probably not the one that Github displays here...

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: You are talking about the DeepRecursion test, right?

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: After implementing the subtree check I will run all the excluded ones again and move if not failing.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 2 above)
Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: Up to you, but I could turn this on its head and say that if we never throw it, it doesn't matter enough to change the type!

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 3 above)
Comment (danmosemsft) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: Why not keep it, tests ideally shouldn't assume the choice of implementation

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: Sorry, the code line that Github displays here seems off. Are you talking about the TODO? If yes then my answer is yes :)

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs [hunk 6]
```diff
@@ -1143,7 +1143,9 @@ private static string SetFromProperty(string capname, bool invert, string patter
                     }
                 }
             }
-            throw new ArgumentException(SR.Format(SR.MakeException, pattern, SR.Format(SR.UnknownProperty, capname)));
+
+            throw new RegexParseException(RegexParseError.UnknownProperty,
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs]: Makes sense. That information should already be accessible in some of the private fields in the RegexParser therefore adding them in the custom exception should be straightforward.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: I just read through the code and both places where the internal errors are thrown can never be reached unless someone changes the detection (the helper functions) of what a quantifier or a special character is. We shouldn't bother too much with that.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.RegularExpressions
+{
+    internal class RegexParseException : ArgumentException
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: This should be [Serializable] right? It's not something that's inherited, so without that, you're breaking serializing the exceptoins from the parser.

And you'll need to add that test case ..

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: Good catch, thanks! And I also need to mask it as an ArgumentException via SetType to not break deserializability on netfx...

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: Note: my tests only cared about ArgumentException.  The other exceptions were catching the internal parser errors that i reported and dan previously fixed :)

The enum would certainly be helpful, to ensure that the expected error is happening, and it's not some other 'argument' issue.  However, as this enum is internal, that won't really help out roslyn that much :)   Given that, i have no horse in this race since this choice is entirely internal and just about helping *you* test things.

If you do end up making these public, i would have a lot more feedback on things.

[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: Right, the current goal is to enrich the current test suite with your tests. At a later point we can definitely talk about exposing more diagnostics to converge different parser implementations.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 8]
```diff
@@ -2286,11 +2289,11 @@ private void PopKeepOptions()
...
 
         /*
-         * Fills in an ArgumentException
+         * Fills in a RegexParseException
          */
-        private ArgumentException MakeException(string message)
+        private RegexParseException MakeException(RegexParseError error, string message)
         {
-            return new ArgumentException(SR.Format(SR.MakeException, _pattern, message));
+            return new RegexParseException(error,_currentPos, SR.Format(SR.MakeException, _pattern, message));
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: nit spacing

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+        /// <summary>
+        /// The error that happened during parsing.
+        /// </summary>
+        public RegexParseError Error => _error;
+
+        /// <summary>
+        /// The offset in the supplied pattern.
+        /// </summary>
+        public int Position { get; }
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: Since users currently can't get to this (without reflection) maybe include it in the message?

**Comment by ViktorHofer** [comment]
Tests are now all passing (except the ones I disabled for netfx 😁). Now I will tackle the remaining two tasks.

**Comment by ViktorHofer** [comment]
Finished with the planned work here. PTAL

**Comment by ViktorHofer** [comment]
so you think the substring manipulation doesn't bring any value here?

**Comment by CyrusNajmabadi** [comment]
The substring manipulations are *very* valuable (they uncovered most of the bugs i found in the .net parser).  However, my point is simply, the roslyn tests use strings of the form:  ```@"..."```

this is because we're generating a c# literal token first and then operating on that. You don't need/want to do that.  You want the contents *inside* that string.

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: I agree and believe we should do at least a one-shot fuzzing over our existing engine.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+
+namespace System.Text.RegularExpressions
+{
+    [Serializable]
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: Do we have a test that breaks when this attribute (or SetType below) is missing?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 11]
```diff
@@ -555,7 +558,7 @@ private RegexCharClass ScanCharClass(bool caseInsensitive, bool scanOnly)
                             if (!scanOnly)
                             {
                                 if (inRange)
-                                    throw MakeException(SR.Format(SR.BadClassInCharRange, ch.ToString()));
+                                    throw MakeException(RegexParseError.BadClassInCharRange, SR.Format(SR.BadClassInCharRange, ch.ToString()));
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs]: Doesn't matter right now, but you could add an overload that takes char, and remove the ch.ToString()'s making slightly smalelr code

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: Note: if you have an actual fuzzer that .net uses, let me know.  If you create a test suite, i would lke to run it against the roslyn parser.  thanks!

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 12]
```diff
@@ -0,0 +1,1095 @@
...
+        [InlineData(@"""  """, RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@"""(?#)""", RegexOptions.None, null, true)]
+        [InlineData(@"""(?# )""", RegexOptions.None, null, true)]
+        [InlineData(@"""(?#""", RegexOptions.None, RegexParseError.UnterminatedComment, true)]
+        [InlineData(@"""(?# """, RegexOptions.None, RegexParseError.UnterminatedComment, true)]
+        [InlineData(@"""(?#)(?#)""", RegexOptions.None, null, true)]
+        [InlineData(@"""(?#)(?#)""", RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@"""(?#) (?#)""", RegexOptions.None, null, true)]
+        [InlineData(@"""(?#) (?#)""", RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@"@""[a\p{Lu}(?#)b]""", RegexOptions.None, null, true)]
```
Comment (danmosemsft) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: What are the nested @ for? I wasn't aware it was special in regexes - but maybe I'm wrong as I see `char ch = '@'; // nonspecial ch, means at beginning` in the code.

Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: The nested ```@""...""``` should be removed.   This is a roslyn-test-ism since i'm first generating a C# string token, since the entire Roslyn embedded-regex stack starts with a String-token which then gets analyzed.  I probably should have written my tests to not have that code and to instead just have the regex contents . I could have wrapped those tests in the test harness with ```@"..."``` instead.  But i wasn't thinking about these being reused here.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: @CyrusNajmabadi now I feel bad that you have to explain several times. When Dan asked I was nearly done changing the test data 😄

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 13]
```diff
@@ -0,0 +1,1076 @@
...
+        [InlineData(@"""|""", RegexOptions.None, null, true)]
+        [InlineData(@""" |""", RegexOptions.None, null, true)]
+        [InlineData(@"""| """, RegexOptions.None, null, true)]
+        [InlineData(@""" | """, RegexOptions.None, null, true)]
+        [InlineData(@"""|""", RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@""" |""", RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@"""| """, RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@""" | """, RegexOptions.IgnorePatternWhitespace, null, true)]
+        [InlineData(@"""||""", RegexOptions.None, null, true)]
+        [InlineData(@"""(|)""", RegexOptions.None, null, true)]
```
Comment (ViktorHofer) [inline src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: @CyrusNajmabadi  Should the quotes from these strings also be trimmed? i.e. `@"""(|)"""` is `"(|)"` but should probably just be `(|)`?

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: I feel bad i put you through this!  It was a silly shortcut i didn't think about when writing these tests.

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: yes.  the test should literally be:  ```[InlineData(@"(|)", ...]```

File: src/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 12 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/tests/RegexParserTests.cs]: Don't feel bad :) You helped us a lot with the test cases. And I know that you are waiting for a response in the other thread, I'll answer as soon as possible.

**Comment by ViktorHofer** [comment]
The error message is now quite nice: `System.Text.RegularExpressions.RegexParseException : Invalid pattern '^[abcd]{1}?*$' at offset 12. Nested quantifier *.`

**Comment by CyrusNajmabadi** [comment]
Will these error messages make it to the full .net framework?  It's a design goal of the Roslyn parser to match your error messages *very* closely.  This is so that the user does see very different message when they actualy run the real regex lib, versus the errros the IDE will show them directly in source.  

So i'll want to update roslyn to follow this.  But it's unclear to me what the story is with corefx changes, and when/how they make it into the desktop.

**Comment by danmosemsft** [comment]
That is nice although strictly there should be single quotes around the quantifier. Probably this one as well     <value>Reference to undefined group name {0}.</value>

Not suggesting you reset your PR..

**Comment by danmosemsft** [comment]
@CyrusNajmabadi there are no plans to port this to .NET Framework.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi there are no plans to port this to .NET Framework.

#sadpandaface

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs (hunk 10 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParseException.cs]: I just added one.

**Comment by ViktorHofer** [comment]
Status: I'm currently trying to repro the error on various machines...

**Comment by CyrusNajmabadi** [comment]
@ViktorHofer Any thoughts about the probable bug in character class parsing htat i pulled you into the discussion on?
